addons/zylann.hterrain/native/src/quad_tree_lod.cpp

```diff
@@ -0,0 +1,219 @@
+    delete _tree;
+}
+
+void QuadTreeLod::_register_methods() {
+	register_method("set_callbacks", &QuadTreeLod::set_callbacks);
+	register_method("clear", &QuadTreeLod::clear);
+	register_method("compute_lod_count", &QuadTreeLod::compute_lod_count);
+	register_method("get_lod_count", &QuadTreeLod::get_lod_count);
+	register_method("get_lod_factor", &QuadTreeLod::get_lod_factor);
+	register_method("create_from_sizes", &QuadTreeLod::create_from_sizes);
+	register_method("set_split_scale", &QuadTreeLod::set_split_scale);
+	register_method("get_split_scale", &QuadTreeLod::get_split_scale);
+	register_method("update", &QuadTreeLod::update);
+	register_method("debug_draw_tree", &QuadTreeLod::debug_draw_tree);
+}
+
+void QuadTreeLod::_init() {
+    Godot::print("QuadTreeLod: _init()");
+}
+
+void QuadTreeLod::set_callbacks(Ref<FuncRef> make_cb, Ref<FuncRef> recycle_cb, Ref<FuncRef> vbounds_cb) {
+    Godot::print("set_callbacks: make_cb->get_function: " + make_cb->get_function());
+	_make_func = make_cb;
+	_recycle_func = recycle_cb;
+	_vertical_bounds_func = vbounds_cb;
+}
+
+void QuadTreeLod::clear() {
+	_join_all_recursively(_tree, _max_depth);
+	_max_depth = 0;
+	_base_size = 0;
+}
+
+int QuadTreeLod::compute_lod_count(int base_size, int full_size) {
+	int po = 0;
+	while (full_size > base_size) {
+		full_size = full_size >> 1;
+		po += 1;
+	}
+	return po;
+}
+
+int QuadTreeLod::get_lod_count() {
+	// TODO make this a count, not max
+	return _max_depth + 1;
+}
+
+int QuadTreeLod::get_lod_factor(int lod) {
+	return 1 << lod;
+}
+
+void QuadTreeLod::create_from_sizes(int base_size, int full_size) {
+	clear();
+	_base_size = base_size;
+	_max_depth = compute_lod_count(base_size, full_size);
+}
+
+// The higher, the longer LODs will spread and higher the quality.
+// The lower, the shorter LODs will spread and lower the quality.
+void QuadTreeLod::set_split_scale(real_t p_split_scale) {
+	real_t MIN = 2.0;
+	real_t MAX = 5.0;
+
+	// Split scale must be greater than a threshold,
+	// otherwise lods will decimate too fast and it will look messy
+	if (p_split_scale < MIN)
+		p_split_scale = MIN;
+	if (p_split_scale > MAX)
+		p_split_scale = MAX;
+
+	_split_scale = p_split_scale;
+}
+
+real_t QuadTreeLod::get_split_scale() {
+	return _split_scale;
+}
+
+void QuadTreeLod::update(Vector3 view_pos) {
+	_update(_tree, _max_depth, view_pos);
+
+	// This makes sure we keep seeing the lowest LOD,
+	// if the tree is cleared while we are far away
+	if (! _tree->has_children() && _tree->data.get_type() == Variant::NIL)
+		_tree->data = _make_chunk(_max_depth, 0, 0);
+}
+
+void QuadTreeLod::_update(Quad *quad, int lod, Vector3 view_pos) {
+    // This function should be called regularly over frames.
+
+    int lod_factor = get_lod_factor(lod);
+    int chunk_size = _base_size * lod_factor;
+    Vector3 world_center = (real_t)chunk_size * (Vector3((real_t)quad->origin_x, 0, (real_t)quad->origin_y) + Vector3(0.5, 0, 0.5));
```

Comment:
Perhaps it is not necessary to cast so explicitely in the C-style way, at least in the parameters to Vector3

Reply:
Every (real_t) cast I put in was just to get rid of compiler warnings. All of these variables are ints being passed to real_t functions.

Reply:
I wonder where these warnings come from. I never get these when I pass ints to functions expecting floats when working on the engine 🤔

Reply:
Your SConstruct lists -W3 which includes `lint-like warnings`.
Godot's sconstruct defaults to no warnings, /W2 if `moderate` is requested, /W3 for `all` and /Wall for `extra`